modules/core/src/main/java/org/apache/ignite/internal/processors/metastorage/persistence/DistributedMetaStorageImpl.java

```diff
@@ -1033,14 +1058,29 @@ else if (!isClient && ver.id() > 0) {
      * @throws IgniteCheckedException If there was an error while sending discovery message.
      */
     private GridFutureAdapter<?> startWrite(String key, byte[] valBytes) throws IgniteCheckedException {
-       if (!isSupported(ctx))
-            throw new IgniteCheckedException(NOT_SUPPORTED_MSG);
+        GridFutureAdapter<?> validationRes = validateBeforeWrite(key);
+
+        if (validationRes != null)
+            return validationRes;
 
         UUID reqId = UUID.randomUUID();
 
         GridFutureAdapter<Boolean> fut = new GridFutureAdapter<>();
 
-        updateFuts.put(reqId, fut);
+        updateFutsStopLock.readLock().lock();
+
+        try {
+            if (stopped) {
+                fut.onDone(nodeStoppingException());
+
```

Comment:
Does it make sense to move this into 'validateBeforeWrite' ?
'validateBeforeWrite' can be executed under read-lock as in normal case it always return null.

Reply:
Yes it does, thank you!